components/private_channel/rust/ffi/private-channel/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+extern crate curve25519_dalek;
+extern crate sha2;
+
+use curve25519_dalek::scalar::Scalar;
+use elgamal_ristretto::ciphertext::Ciphertext;
+use elgamal_ristretto::private::SecretKey;
+use repsys_crypto::{
+    combine_pks, encrypt_input, generate_keys, partial_decryption_and_proof, randomize_and_prove,
+};
+
+use sha2::Sha512;
+use std::vec::Vec;
+
+use bincode;
+
+pub fn start_challenge(
+    input: Vec<String>,
+    server_pk_encoded: [u8; 32],
+) -> Result<(Vec<u8>, Vec<u8>, Vec<u8>, Vec<u8>), ()> {
+    let pk_server = match bincode::deserialize(&server_pk_encoded) {
+			Err(_) => return Err(()),
+			Ok(pk) => pk,
+		};
+
+    let mut vector_hashes: Vec<Scalar> = Vec::new();
+    for n in 0..input.len() {
+        vector_hashes.push(Scalar::hash_from_bytes::<Sha512>(input[n].as_bytes()));
+    }
+
+    // generate new key pair at every round
+    let (sk_client, pk_client) = generate_keys();
+    let shared_pk = combine_pks(pk_client, pk_server);
+
+    let encrypted_hashes = encrypt_input(shared_pk, &vector_hashes);
+
+    let encoded_pk = match bincode::serialize(&pk_client) {
+      Err(_) => return Err(()),
+      Ok(pk) => pk,
+    };
```

Comment:
Good opportunities to simplify the code with the `?` operator here and below:

```rust
let encoded_pk = match bincode::serialize(&pk_client).map_err(|_| ())?;
```

Reply:
- [ ] simplify code with `map_err`